cpp/nccl_comms/src/cuML_comms_nccl_impl.cpp

```diff
@@ -0,0 +1,445 @@
+    case MLCommon::cumlCommunicator::UINT8:
+      return sizeof(unsigned char);
+    case MLCommon::cumlCommunicator::INT:
+      return sizeof(int);
+    case MLCommon::cumlCommunicator::UINT:
+      return sizeof(unsigned int);
+    case MLCommon::cumlCommunicator::INT64:
+      return sizeof(long long int);
+    case MLCommon::cumlCommunicator::UINT64:
+      return sizeof(unsigned long long int);
+    case MLCommon::cumlCommunicator::FLOAT:
+      return sizeof(float);
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(
+  const cumlNCCLCommunicator_impl::datatype_t datatype) {
+  switch (datatype) {
+    case MLCommon::cumlCommunicator::CHAR:
+      return ncclChar;
+    case MLCommon::cumlCommunicator::UINT8:
+      return ncclUint8;
+    case MLCommon::cumlCommunicator::INT:
+      return ncclInt;
+    case MLCommon::cumlCommunicator::UINT:
+      return ncclUint32;
+    case MLCommon::cumlCommunicator::INT64:
+      return ncclInt64;
+    case MLCommon::cumlCommunicator::UINT64:
+      return ncclUint64;
+    case MLCommon::cumlCommunicator::FLOAT:
+      return ncclFloat;
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const cumlNCCLCommunicator_impl::op_t op) {
+  switch (op) {
+    case MLCommon::cumlCommunicator::SUM:
+      return ncclSum;
+    case MLCommon::cumlCommunicator::PROD:
+      return ncclProd;
+    case MLCommon::cumlCommunicator::MIN:
+      return ncclMin;
+    case MLCommon::cumlCommunicator::MAX:
+      return ncclMax;
+  }
+}
+}  // namespace
+
+static const ucp_tag_t default_tag_mask = -1;
+
+static void wait(ucp_worker_h ucp_worker, struct ucx_context *context) {
+  while (context->completed == 0) {
+    ucp_worker_progress(ucp_worker);
+  }
+}
+
+/**
+ * Underlying comms, like NCCL and UCX, should be initialized and ready for use
+ * outside of the cuML Comms lifecycle. This allows us to decouple the ownership
+ * of the actual comms from cuml so that they can also be used directly, outside of
+ * cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python as well, before being used to construct a cuML comms instance.
+ */
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  ucp_ep_h *eps, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlNCCLCommunicator_impl(comm, ucp_worker, eps, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+
+void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm, void *ucp_worker,
+                     void *eps, int size, int rank) {
+  ucp_worker_print_info((ucp_worker_h)ucp_worker, stdout);
+
+  ucp_ep_h *new_ep_arr = new ucp_ep_h[size];
```

Comment:
Who owns the memory of `new_ep_arr` and is responsible for freeing it?

Reply:
Correct, I will tag people once ready to review. There's going to be a lot of little places currently where little things like this occur because of the movement of functions around right now. These should all be cleaned up once I'm ready for review.